### ISSUE TITLE
System Hard Lock on Kernel 6.4.x on Fedora 38

### ISSUE THREAD

**Issue by JustPlainGarak** [LizardByte/Sunshine#1464]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your issue described in the documentation?

- [X] I have read the documentation

### Is your issue present in the nightly release?

- [X] This issue is present in the nightly release

### Describe the Bug

When I launch any game from a Sunshine client connected to my Fedora 38 host while running kernel version 6.4.x on the host, the host computer soft locks upon game render starting. The client will remain connected for a few seconds on a frozen screen and then boot me out. When I arrive physically in front of the host, I find it soft locked on the last image I saw from the client and must physically power cycle the host. I am currently running kernel 6.4.4-200.fc38.x86_64 on the host when this issue occurs, though the issue occurs on every version kernel 6.4.x that I have tried.

### Expected Behavior

The game render should run normally and not cause a soft lock of my host OS.

### Additional Context

The issue does not occur when my host PC is running any version of the 6.3.x kernel in Fedora 38 (so I have a workaround for now). I can, without issue, run any game I want on kernel 6.3.12-200.fc38.x86_64. With no other changes beyond rebooting to the 6.4.4-200.fc38.x86_64 kernel, a soft lock upon game render starting.

I have tried many different kernels, mesa driver and vaapi driver combinations (ensuring I have vaapi drivers that include h264_vaapi and hevc_vaapi) and the only configuration that produces the soft lock, regardless of other factors, is running a kernel version 6.4.x.

The issue does not occur when my host PC is running Windows.

### Host Operating System

Linux

### Operating System Version

Fedora Linux 38 (KDE Plasma)

### Architecture

64 bit

### Sunshine commit or version

0.20.0.4b986b26c2c921b57b4ab300b002bb9e97e14c93

### Package

Linux - rpm

### GPU Type

AMD

### GPU Model

AMD Radeon RX 7900 XTX

### GPU Driver/Mesa Version

23.1.3

### Capture Method (Linux Only)

KMS

### Config

```Shell
sunshine_name = EmpokNor
encoder = vaapi
amd_rc = auto
hevc_mode = 0
origin_pin_allowed = lan
key_rightalt_to_key_win = enabled
resolutions = [
    352x240,
    480x360,
    858x480,
    1280x720,
    1920x1080,
    2560x1080,
    3440x1440,
    1920x1200,
    3860x2160,
    3840x1600,
    2560x1440
]
min_log_level = 2
nv_rc = auto
capture = kms
nv_preset = default
min_threads = 2
amd_quality = default
```

### Relevant log output

```shell
tail: .config/sunshine/sunshine.log: file truncated
[2023:07:23:22:14:22]: Info: Sunshine version: 0.20.0.4b986b26c2c921b57b4ab300b002bb9e97e14c93
[2023:07:23:22:14:22]: Info: Found display [wayland-0]
[2023:07:23:22:14:22]: Info: Found display [wayland-0]
[2023:07:23:22:14:22]: Info: Found interface: zxdg_output_manager_v1(32) version 3
[2023:07:23:22:14:22]: Info: Found interface: wl_output(50) version 4
[2023:07:23:22:14:22]: Info: Offset: 0x0
[2023:07:23:22:14:22]: Info: Resolution: 2560x1440
[2023:07:23:22:14:22]: Info: Name: DP-1
[2023:07:23:22:14:22]: Info: Found monitor: Samsung Electric Company LC32G7xT/H4ZN801004
[2023:07:23:22:14:22]: Info: DP-1: Samsung Electric Company LC32G7xT/H4ZN801004
[2023:07:23:22:14:22]: Info: Trying encoder [vaapi]
[2023:07:23:22:14:22]: Info: Screencasting with KMS
[2023:07:23:22:14:22]: Info: Found monitor for DRM screencasting
[2023:07:23:22:14:22]: Info: System tray created
[2023:07:23:22:14:22]: Info: SDR color coding [Rec. 601]
[2023:07:23:22:14:22]: Info: Color range: [JPEG]
[2023:07:23:22:14:22]: Info: Screencasting with KMS
[2023:07:23:22:14:22]: Info: Found monitor for DRM screencasting
[2023:07:23:22:14:22]: Info: SDR color coding [Rec. 601]
[2023:07:23:22:14:22]: Info: Color range: [JPEG]
[2023:07:23:22:14:22]: Info: Screencasting with KMS
[2023:07:23:22:14:22]: Info: Found monitor for DRM screencasting
[2023:07:23:22:14:22]: Info: SDR color coding [Rec. 709]
[2023:07:23:22:14:22]: Info: Color range: [JPEG]
[2023:07:23:22:14:22]: Warning: vaapi: hevc missing sps->vui parameters
[2023:07:23:22:14:22]: Info: // Testing for available encoders, this may generate errors. You can safely ignore those errors. //
[2023:07:23:22:14:22]: Info: 
[2023:07:23:22:14:22]: Info: // Ignore any errors mentioned above, they are not relevant. //
[2023:07:23:22:14:22]: Info: 
[2023:07:23:22:14:22]: Info: Found encoder vaapi: [h264_vaapi, hevc_vaapi]
[2023:07:23:22:14:22]: Info: Adding avahi service Sunshine
[2023:07:23:22:14:22]: Info: Configuration UI available at [https://localhost:47990]
[2023:07:23:22:14:23]: Info: Avahi service Sunshine successfully established.
[2023:07:23:22:14:31]: Info: Trying encoder [vaapi]
[2023:07:23:22:14:31]: Info: Screencasting with KMS
[2023:07:23:22:14:31]: Info: Found monitor for DRM screencasting
[2023:07:23:22:14:31]: Info: SDR color coding [Rec. 601]
[2023:07:23:22:14:31]: Info: Color range: [JPEG]
[2023:07:23:22:14:31]: Info: Screencasting with KMS
[2023:07:23:22:14:31]: Info: Found monitor for DRM screencasting
[2023:07:23:22:14:31]: Info: SDR color coding [Rec. 601]
[2023:07:23:22:14:31]: Info: Color range: [JPEG]
[2023:07:23:22:14:31]: Info: Screencasting with KMS
[2023:07:23:22:14:31]: Info: Found monitor for DRM screencasting
[2023:07:23:22:14:31]: Info: SDR color coding [Rec. 709]
[2023:07:23:22:14:31]: Info: Color range: [JPEG]
[2023:07:23:22:14:31]: Warning: vaapi: hevc missing sps->vui parameters
[2023:07:23:22:14:31]: Info: // Testing for available encoders, this may generate errors. You can safely ignore those errors. //
[2023:07:23:22:14:31]: Info: 
[2023:07:23:22:14:31]: Info: // Ignore any errors mentioned above, they are not relevant. //
[2023:07:23:22:14:31]: Info: 
[2023:07:23:22:14:31]: Info: Found encoder vaapi: [h264_vaapi, hevc_vaapi]
[2023:07:23:22:14:31]: Info: Executing [Desktop]
[2023:07:23:22:14:31]: Info: CLIENT CONNECTED
[2023:07:23:22:14:31]: Info: Found display [wayland-0]
[2023:07:23:22:14:31]: Info: Found interface: zxdg_output_manager_v1(32) version 3
[2023:07:23:22:14:31]: Info: Found interface: wl_output(50) version 4
[2023:07:23:22:14:31]: Info: Offset: 0x0
[2023:07:23:22:14:31]: Info: Resolution: 2560x1440
[2023:07:23:22:14:31]: Info: Name: DP-1
[2023:07:23:22:14:31]: Info: Found monitor: Samsung Electric Company LC32G7xT/H4ZN801004
[2023:07:23:22:14:31]: Info: DP-1: Samsung Electric Company LC32G7xT/H4ZN801004
[2023:07:23:22:14:31]: Info: Screencasting with KMS
[2023:07:23:22:14:31]: Info: Found monitor for DRM screencasting
[2023:07:23:22:14:31]: Info: SDR color coding [Rec. 601]
[2023:07:23:22:14:31]: Info: Color range: [MPEG]
[2023:07:23:22:14:32]: Info: Setting default sink to: [sink-sunshine-stereo]
[2023:07:23:22:14:32]: Info: Found default monitor by name: sink-sunshine-stereo.monitor
[2023:07:23:22:15:00]: Info: CLIENT DISCONNECTED
[2023:07:23:22:15:00]: Info: Setting default sink to: [alsa_output.usb-Creative_Technology_Ltd_Sound_Blaster_X3_9CE2D6EF69D9BF0A-03.analog-stereo]

Message from syslogd@EmpokNor at Jul 23 22:15:15 ...
 kernel:watchdog: Watchdog detected hard LOCKUP on cpu 12

Message from syslogd@EmpokNor at Jul 23 22:15:15 ...
 kernel:watchdog: BUG: soft lockup - CPU#6 stuck for 26s! [FSD-Win64-Shipp:10436]

Message from syslogd@EmpokNor at Jul 23 22:15:15 ...
 kernel:watchdog: BUG: soft lockup - CPU#8 stuck for 26s! [steamwebhelper:6958]

Message from syslogd@EmpokNor at Jul 23 22:15:15 ...
 kernel:watchdog: BUG: soft lockup - CPU#13 stuck for 22s! [kworker/13:1:9113]

Message from syslogd@EmpokNor at Jul 23 22:15:23 ...
 kernel:watchdog: BUG: soft lockup - CPU#10 stuck for 22s! [kworker/10:0:9241]

Message from syslogd@EmpokNor at Jul 23 22:15:43 ...
 kernel:watchdog: BUG: soft lockup - CPU#6 stuck for 52s! [FSD-Win64-Shipp:10436]

Message from syslogd@EmpokNor at Jul 23 22:15:43 ...
 kernel:watchdog: BUG: soft lockup - CPU#8 stuck for 52s! [steamwebhelper:6958]

Message from syslogd@EmpokNor at Jul 23 22:15:43 ...
 kernel:watchdog: BUG: soft lockup - CPU#13 stuck for 48s! [kworker/13:1:9113]

Message from syslogd@EmpokNor at Jul 23 22:15:51 ...
 kernel:watchdog: BUG: soft lockup - CPU#10 stuck for 48s! [kworker/10:0:9241]

Message from syslogd@EmpokNor at Jul 23 22:16:11 ...
 kernel:watchdog: BUG: soft lockup - CPU#8 stuck for 78s! [steamwebhelper:6958]

Message from syslogd@EmpokNor at Jul 23 22:16:11 ...
 kernel:watchdog: BUG: soft lockup - CPU#13 stuck for 74s! [kworker/13:1:9113]

Message from syslogd@EmpokNor at Jul 23 22:16:15 ...
 kernel:watchdog: BUG: soft lockup - CPU#6 stuck for 82s! [FSD-Win64-Shipp:10436]

Message from syslogd@EmpokNor at Jul 23 22:16:19 ...
 kernel:watchdog: BUG: soft lockup - CPU#10 stuck for 74s! [kworker/10:0:9241]

Message from syslogd@EmpokNor at Jul 23 22:16:39 ...
 kernel:watchdog: BUG: soft lockup - CPU#8 stuck for 104s! [steamwebhelper:6958]

Message from syslogd@EmpokNor at Jul 23 22:16:39 ...
 kernel:watchdog: BUG: soft lockup - CPU#13 stuck for 100s! [kworker/13:1:9113]
client_loop: send disconnect: Broken pipe
```

**Comment by cgutman** [comment]
This is definitely a hardware or kernel issue. A regular userspace program cannot directly cause a soft lockup like this (that would be a denial of service vulnerability).

Check your `dmesg` output to see if there's a backtrace for the stuck CPUs.

**Comment by CristianPi** [comment]
fedora 38 6.4.4-200.fc38.x86_64, gnome xorg, mesa locally built (with hw codecs enabled) works, 6800.

**Comment by JustPlainGarak** [comment]
> This is definitely a hardware or kernel issue. A regular userspace program cannot directly cause a soft lockup like this (that would be a denial of service vulnerability).
> 
> Check your `dmesg` output to see if there's a backtrace for the stuck CPUs.

Yeah, its VERY strange that the cutoff is so stark from 6.3 and 6.4. I'm also running Wayland, I may go back and see if I have the problem in X11 as well (with KMS and X11 capture). I've attached some dmesg captures taken, the kernel used, whether I was using Sunshine or not, and whether I was able to launch a game without hanging are specified in the filenames.
[dmesg_6.3.12_sunshine_success.log](https://github.com/LizardByte/Sunshine/files/12167542/dmesg_6.3.12_sunshine_success.log)
[dmesg_6.4.4_console_success.log](https://github.com/LizardByte/Sunshine/files/12167544/dmesg_6.4.4_console_success.log)
[dmesg_6.4.4_sunshine_failure.log](https://github.com/LizardByte/Sunshine/files/12167545/dmesg_6.4.4_sunshine_failure.log)

**Comment by JustPlainGarak** [comment]
I strongly suspect that the issue is related to possible changes to how KMS screen capture works in kernel 6.4, but I am far from a developer and cannot parse what may or may not have changed on that front.

I cannot test capturing with wlroots because I am running KDE Plasma and kwin pipewire capture isn't supported by Sunshine (yet). I cannot test X11 capture because my system doesn't initialize a desktop when I log in with X11.

Would it be possible for someone on kernel 6.4.x who normally uses X11 to login to a Wayland desktop and test with KMS capture to see if anyone can at least duplicate what I'm seeing here?

I'll also attach a journalctl log of a hung system I took if anyone looking at this is skilled with parsing these logs and can interpret what is happening to cause these lockups.
[journalctl_hang_2023-08-09.log](https://github.com/LizardByte/Sunshine/files/12328310/journalctl_hang_2023-08-09.log)

**Comment by JustPlainGarak** [comment]
Whatever the issue was in kernel 6.4.x, it appears to be resolved on Fedora since kernel 6.5.5.
